Add cert manager migration docs (#179)
openshift-service-mesh/sail-operator: docs/ossm/ossm2-migration/cert-manager/README.md

spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca
  namespace: cert-manager
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
```

1. Install istio-ca

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  isCA: true
  duration: 21600h
  secretName: istio-ca
  commonName: istio-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  ca:
    secretName: istio-ca
```

1. Helm install istio-csr

```console
helm repo add jetstack https://charts.jetstack.io --force-update
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
    --install \

>     --namespace cert-manager \

Author: iirc if you use `istio-system` you have to `--set` more options when installing istio-csr. iirc when I originally wrote the cert-manager docs I was using the [upstream docs](https://cert-manager.io/docs/usage/istio-csr/installation/#5-installing-istio-csr) which has you install istio-csr in the `cert-manager` namespace so that's why it's different.

>     --namespace cert-manager \

Author: The 3.0 documentation uses `cert-manager`: https://docs.redhat.com/en/documentation/red_hat_openshift_service_mesh/3.0.0tp1/html/installing/ossm-cert-manager-assembly#inplace-istio-csr-installation_ossm-cert-manager-assembly so it's a mixed bag. I can update the namespace here to be `istio-system` we'll also need to update the existing 3.0 docs if we want everything to match.

```
 -->

Starting with a `ServiceMeshControlPlane` with cert-manager configured:

```yaml
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  ...
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.cert-manager.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
```

You will need to perform these updates to your istio-csr deployment:

- The `app.istio.revisions` field needs to include your 3.0 control plane revision _before_ you create your `Istio` resource.

  Adding your 3.0 control plane revision to your istio-csr deployment will ensure that proxies can properly communicate with the 3.0 control plane.

  ```console
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace cert-manager \
      --wait \
      --set "app.istio.revisions={basic,ossm3-v1-24-1}"
  ```

>   Depending on whether you will use a `RevisionBased` update strategy or an `InPlace` update strategy, your revision name will vary. If using an `InPlace` strategy, your revision name will match your `Istio` name. If using a `RevisionBased` strategy, revision names use the following format, <istio-name>-v<major_version>-<minor_version>-<patch_version>. For example: `ossm3-v1-24-1`.

Author: Ah you're right. Must be some markdown processing error.

- The `app.controller.configmapNamespaceSelector` field needs to be either unset _before_ the migration begins or updated _after_ you have completed your migration.

>   If you have set the `app.controller.configmapNamespaceSelector` field on your istio-csr deployment to `maistra.io/member-of`, you will need to update this accordingly. If you haven't set this field, you can keep it unset.

Reviewer: Is it possible to have multiple labels defined in `app.controller.configmapNamespaceSelector`? It would be nice to include also some example when you need to keep `maistra.io/member-of` so 2.x is still working while 3.0 namespace would use a different label.

Author: >It would be nice to include also some example when you need to keep maistra.io/member-of so 2.x is still working while 3.0 namespace would use a different label.

This is where you need to update the selector **only after** completing the migration and have also added the new injection label to each namespace.

Author: I'll reword it to make it more clear that you don't always need the injection label.

- The `app.controller.configmapNamespaceSelector` field needs to be either unset _before_ the migration begins or updated _after_ you have completed your migration.

  If you have set the `app.controller.configmapNamespaceSelector` field on your istio-csr deployment to `maistra.io/member-of`, you will need to update this accordingly. If you haven't set this field, you can keep it unset.

  If unset the `configmapNamespaceSelector` field on your istio-csr deployment, the istio CA configmap will be injected into every namespace. `MultiTenant` deployments with more than one `ServiceMeshControlPlane` in the cluster should not remove this field since the wrong CA configmap would likely get written to the namespace. If you are keeping the field set, you need to wait until **after** you have completed your migration to update the `configmapNamespaceSelector` field. Otherwise namespaces without the injection label will no longer have the configmap CA injected.

  To unset this field:

  ```console
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace cert-manager \
      --wait \

>       --set "app.controller.configmapNamespaceSelector=null"

Reviewer: This was not working for me, the value was unchanged after this. I used `--set "app.controller.configmapNamespaceSelector="`.

Author: I must've copied an array example that used `null`

kind: Issuer
metadata:
  name: selfsigned-root-issuer
  namespace: cert-manager
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca
  namespace: cert-manager
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
```

1. Install istio-ca

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  isCA: true
  duration: 21600h
  secretName: istio-ca
  commonName: istio-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  ca:
    secretName: istio-ca
```

1. Helm install istio-csr

```console
helm repo add jetstack https://charts.jetstack.io --force-update

> helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \

Reviewer: Sorry for complication but since this is 2.6 -> 3.0 migration guide, why not just following the 2.6 OSSM [documentation](https://docs.openshift.com/container-platform/4.16/service_mesh/v2x/ossm-security.html#ossm-cert-manager-installation_ossm-security) as a starting point? That way we will have the same starting point as our customers.

Author: That makes sense. I was following the [3.0 OSSM docs for cert-manager](https://docs.redhat.com/en/documentation/red_hat_openshift_service_mesh/3.0.0tp1/html/installing/ossm-cert-manager-assembly#ossm-installing-cert-manager_ossm-cert-manager-assembly).

    --install \
    --namespace istio-system \
    --wait \
    --set "app.tls.rootCAFile=/var/run/secrets/istio-csr/ca.pem" \
    --set "volumeMounts[0].name=root-ca" \
    --set "volumeMounts[0].mountPath=/var/run/secrets/istio-csr" \
    --set "volumes[0].name=root-ca" \
    --set "volumes[0].secret.secretName=istio-ca" \
    --set "volumes[0].secret.items[0].key=tls.crt" \
    --set "volumes[0].secret.items[0].path=ca.pem" \
    --set "app.istio.revisions={basic}"
```

1. Create SMCP

```yaml
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    grafana:
      enabled: false
    kiali:
      enabled: false
    prometheus:
      enabled: false
  gateways:
    enabled: false
    openshiftRoute:
      enabled: false
  profiles:
    - default
  security:
    certificateAuthority:
      cert-manager:

>         address: cert-manager-istio-csr.cert-manager.svc:443

Reviewer: I think this should be `cert-manager-istio-csr.istio-system.svc:443`

      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
  tracing:
    type: None
  version: v2.6
```

1. Update istio-csr with future Istio revision.

```console
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
    --install \
    --reuse-values \

>     --namespace cert-manager \

Reviewer: ```suggestion
    --namespace istio-system \
```

    --wait \
    --set "app.istio.revisions={basic,ossm3-v1-24-1}"
```

1. Create Istio resource

```yaml
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  generation: 3
  name: ossm3
spec:
  namespace: istio-system
  updateStrategy:
    type: RevisionBased
  values:
    global:

>       caAddress: cert-manager-istio-csr.cert-manager.svc:443

Reviewer: ```suggestion
      caAddress: cert-manager-istio-csr.istio-system.svc:443
```

productpage-v1-8669b4d5c8-hshtz.bookinfo     Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-ossm3-v1-24-1-d5b9b4c89-ccz4v     1.24.1
ratings-v1-bbb89988d-tcgvp.bookinfo          Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-ossm3-v1-24-1-d5b9b4c89-ccz4v     1.24.1
reviews-v1-75b6949cf4-7kbdm.bookinfo         Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-ossm3-v1-24-1-d5b9b4c89-ccz4v     1.24.1
reviews-v2-64f68558b-gsxc4.bookinfo          Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-ossm3-v1-24-1-d5b9b4c89-ccz4v     1.24.1
reviews-v3-596954cfd6-jnb6n.bookinfo         Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-ossm3-v1-24-1-d5b9b4c89-ccz4v     1.24.1
```
 -->

Starting with a `ServiceMeshControlPlane` with cert-manager configured:

```yaml
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  ...
  security:
    certificateAuthority:
      cert-manager:

>         address: cert-manager-istio-csr.cert-manager.svc:443

Reviewer: ```suggestion
        address: cert-manager-istio-csr.istio-system.svc:443
```

  ```console
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace istio-system \
      --wait \
      --set "app.controller.configmapNamespaceSelector="
  ```

  To update this field:

  > **_NOTE:_** Before updating, ensure you have completely finished your migration and the new injection label, in this example `istio-injection=enabled`, is present on all workload namespaces before updating istio-csr.

  ```console
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
     --install \
     --reuse-values \
     --namespace istio-system \
     --wait \
     --set "app.controller.configmapNamespaceSelector=istio-injection=enabled"
  ```

After updating your istio-csr deployment, you can create the `Istio` resource with the following settings to work with cert-manager. Similar to the 2.6 controlplane, these settings disable the built-in CA server and instead use the istio-csr address.

- Create Istio Resource.

  ```yaml
  apiVersion: sailoperator.io/v1alpha1
  kind: Istio
  metadata:
    name: ossm3
  spec:
    ...
    namespace: istio-system
    values:
      global:

>         caAddress: cert-manager-istio-csr.cert-manager.svc:443

Reviewer: ```suggestion
        caAddress: cert-manager-istio-csr.istio-system.svc:443
```

# OpenShift Service Mesh 2.6 --> 3.0 Migration with Cert-Manager

> When migrating from OpenShift Service Mesh 2.6 --> 3.0 while using Cert-Manager you can largely follow the [ClusterWide](TODO) or [MultiTenant](../multi-tenancy/README.md) migration guides with a few extra steps detailed here.

Reviewer 2: ```suggestion
When migrating from OpenShift Service Mesh 2.6 --> 3.0 while using Cert-Manager you can largely follow the [ClusterWide](TODO) or [MultiTenant](../multi-tenancy/README.md) migration guides. This document details a few necessary additional steps to follow before creating your `Istio` resource to ensure your cert manager configuration works with 3.0. 
```
